File: tests/java/tests-java-gradle-junit5-allure3-playwright/src/test/java/helpers/SelenoidPlaywrightEndpoint.java
package helpers;

import java.net.URI;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.LinkedHashMap;
import java.util.Locale;
import java.util.stream.Collectors;

/**
 * Selenoid Playwright is a WebSocket ({@code wss://…/playwright/playwright-chromium/…}),
 * not WebDriver {@code /wd/hub}. Query may contain {@code accessKey} — never log it.
 */
public final class SelenoidPlaywrightEndpoint {

    private SelenoidPlaywrightEndpoint() {
    }

    /**
     * Prefer {@code SELENOID_PLAYWRIGHT_URL} (Jenkins/GHA env keeps {@code ?accessKey=}).
     * {@code -DremoteUrl=wss://…?accessKey=} is easy to truncate before the test JVM.
     */
    public static String resolve(String configRemoteUrl) {
        return preferWebSocket(System.getenv("SELENOID_PLAYWRIGHT_URL"), configRemoteUrl);
    }

    public static String preferWebSocket(String envUrl, String configUrl) {
        var env = envUrl == null ? "" : envUrl.trim();
        if (isWebSocket(env)) {
            return env;
        }
        return configUrl == null ? "" : configUrl.trim();
    }

    public static boolean isWebSocket(String url) {
        if (url == null || url.isBlank()) {
            return false;
        }
        var u = url.trim().toLowerCase(Locale.ROOT);
        return u.startsWith("ws://") || u.startsWith("wss://");
    }

    public static boolean isHttpUrl(String url) {
        if (url == null || url.isBlank()) {
            return false;
        }
        var u = url.trim().toLowerCase(Locale.ROOT);
        return u.startsWith("http://") || u.startsWith("https://");
    }

    /** Scheme + host + path only. */
    public static String describe(String url) {
        if (url == null || url.isBlank()) {
            return "";
        }
        try {
            var uri = URI.create(url.trim());
            var host = uri.getHost() == null ? "" : uri.getHost();
            var path = uri.getPath() == null ? "" : uri.getPath();
            var scheme = uri.getScheme() == null ? "" : uri.getScheme();
            return scheme + "://" + host + path;
        } catch (IllegalArgumentException e) {
            return "(unparseable remoteUrl)";
        }
    }

    public static String withSessionQuery(String ws, boolean enableVnc, boolean enableVideo) {
        return withSessionQuery(ws, enableVnc, enableVideo, null, null);
    }

    /**
     * Playwright has no WebDriver session id. When the hub records video, pass a unique
     * {@code videoName} so Allure can link {@code videoFolder}/{videoName} the same way
     * the TypeScript Playwright cell does.
     */
    public static String withSessionQuery(
            String ws,
            boolean enableVnc,
            boolean enableVideo,
            String videoName,
            String screenResolution) {
        var base = ws.trim();
        var extra = new LinkedHashMap<String, String>();
        extra.put("name", "autotests-ai-multistack-java-pw");
        extra.put("sessionTimeout", "5m");
        extra.put("enableVNC", enableVnc ? "true" : "false");
        extra.put("enableVideo", enableVideo ? "true" : "false");
        if (enableVideo) {
            if (videoName != null && !videoName.isBlank()) {
                extra.put("videoName", videoName.trim());
            }
            if (screenResolution != null && !screenResolution.isBlank()) {
                extra.put("screenResolution", screenResolution.trim());
            }
        }
        var encoded = extra.entrySet().stream()
                .map(e -> enc(e.getKey()) + "=" + enc(e.getValue()))
                .collect(Collectors.joining("&"));
        return base.contains("?") ? base + "&" + encoded : base + "?" + encoded;
    }

    /** Public Selenoid video URL: {@code https://selenoid.qa.guru/video/<file>.mp4}. */
    public static String videoUrl(String folder, String fileName) {
        if (folder == null || folder.isBlank() || fileName == null || fileName.isBlank()) {
            return "";
        }
        var base = folder.trim();
        if (!base.endsWith("/")) {
            base = base + "/";
        }
        return base + fileName.trim();
    }

    private static String enc(String s) {
        return URLEncoder.encode(s, StandardCharsets.UTF_8);
    }
}

File: tests/csharp/tests-csharp-xunit-api_request-playwright/Helpers/SelenoidPlaywrightEndpoint.cs
namespace Helpers;

/// <summary>
/// Selenoid Playwright is a WebSocket (<c>wss://…/playwright/playwright-chromium/…</c>),
/// not WebDriver <c>/wd/hub</c>. Query may contain <c>accessKey</c> — never log it.
/// </summary>
public static class SelenoidPlaywrightEndpoint
{
    /// <summary>
    /// Prefer <c>SELENOID_PLAYWRIGHT_URL</c> (Jenkins/GHA env keeps <c>?accessKey=</c>).
    /// <c>REMOTE_URL=wss://…?accessKey=</c> is easy to truncate before the test host.
    /// </summary>
    public static string Resolve(string configRemoteUrl) =>
        PreferWebSocket(Environment.GetEnvironmentVariable("SELENOID_PLAYWRIGHT_URL"), configRemoteUrl);

    public static string PreferWebSocket(string? envUrl, string? configUrl)
    {
        var env = envUrl?.Trim() ?? "";
        if (IsWebSocket(env))
        {
            return env;
        }

        return configUrl?.Trim() ?? "";
    }

    public static bool IsWebSocket(string? url)
    {
        if (string.IsNullOrWhiteSpace(url))
        {
            return false;
        }

        var u = url.Trim().ToLowerInvariant();
        return u.StartsWith("ws://") || u.StartsWith("wss://");
    }

    public static bool IsHttpUrl(string? url)
    {
        if (string.IsNullOrWhiteSpace(url))
        {
            return false;
        }

        var u = url.Trim().ToLowerInvariant();
        return u.StartsWith("http://") || u.StartsWith("https://");
    }

    /// <summary>Scheme + host + path only.</summary>
    public static string Describe(string? url)
    {
        if (string.IsNullOrWhiteSpace(url))
        {
            return "";
        }

        try
        {
            var uri = new Uri(url.Trim());
            return uri.Scheme + "://" + uri.Host + uri.AbsolutePath;
        }
        catch (UriFormatException)
        {
            return "(unparseable remoteUrl)";
        }
    }

    public static string WithSessionQuery(string ws, bool enableVnc, bool enableVideo) =>
        WithSessionQuery(ws, enableVnc, enableVideo, null, null);

    /// <summary>
    /// Playwright has no WebDriver session id. When the hub records video, pass a unique
    /// <c>videoName</c> so Allure can link <c>videoFolder/videoName</c> the same way
    /// the Java/TypeScript Playwright cells do.
    /// </summary>
    public static string WithSessionQuery(
        string ws,
        bool enableVnc,
        bool enableVideo,
        string? videoName,
        string? screenResolution)
    {
        var extra = new Dictionary<string, string>
        {
            ["name"] = "autotests-ai-multistack-csharp-pw",
            ["sessionTimeout"] = "5m",
            ["enableVNC"] = enableVnc ? "true" : "false",
            ["enableVideo"] = enableVideo ? "true" : "false",
        };
        if (enableVideo)
        {
            if (!string.IsNullOrWhiteSpace(videoName))
            {
                extra["videoName"] = videoName.Trim();
            }

            if (!string.IsNullOrWhiteSpace(screenResolution))
            {
                extra["screenResolution"] = screenResolution.Trim();
            }
        }

        var encoded = string.Join("&", extra.Select(e => Enc(e.Key) + "=" + Enc(e.Value)));
        var baseWs = ws.Trim();
        return baseWs.Contains('?') ? baseWs + "&" + encoded : baseWs + "?" + encoded;
    }

    /// <summary>Public Selenoid video URL: <c>https://selenoid.qa.guru/video/file.mp4</c>.</summary>
    public static string VideoUrl(string? folder, string? fileName)
    {
        if (string.IsNullOrWhiteSpace(folder) || string.IsNullOrWhiteSpace(fileName))
        {
            return "";
        }

        var baseFolder = folder.Trim();
        if (!baseFolder.EndsWith('/'))
        {
            baseFolder += "/";
        }

        return baseFolder + fileName.Trim();
    }

    private static string Enc(string s) => Uri.EscapeDataString(s);
}
